### PR TITLE
Azure: New test case of rescind the VF device from guest

### PR DESCRIPTION
**Pull request by xian123**

### CODE REVIEW

**Comment by xian123** [comment]
Test result:
03/29/2019 08:48:22 : [INFO ]
[LISAv2 Test Results Summary]
Test Run On           : 03/29/2019 08:36:38
ARM Image Under Test  : canonical : ubuntuserver : 18.04-lts : Latest
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:11

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes)
------------------------------------------------------------------------------------------------------------------
    1 SRIOV                SRIOV-HOTPLUG-VF-DEVICE                                                           PASS                 7.24

**Comment by chvalean** [comment]
@xian123 please also test this with RHEL 7.x and SLES15.
Not necessarily to post test results, but more to ensure that the automation works with those distros as well.

**Review by adriansuhov — APPROVED** [review]

**Comment by adriansuhov** [comment]
As @chvalean said, please test RHEL 7.5 and SLES 15. We will merge after

File: Testscripts/Linux/SRIOV-Hotplug-VF-Device.sh [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+LogMsg "Starting iperf in server mode on ${VF_IP2}"
+ssh -i "$HOME"/.ssh/"$SSH_PRIVATE_KEY" -o StrictHostKeyChecking=no ${remote_user}@${VF_IP2} "nohup iperf3 -s > client.out &" > /dev/null 2>&1
+if [ $? -ne 0 ]; then
+    LogErr "Unable to start iperf server script on the dependency vm."
+    SetTestStateFailed
+    exit 0
+fi
+
+cat > Start_Iperf_Client.sh << EOF
+while true
```
Comment (juhlee-microsoft) [inline Testscripts/Linux/SRIOV-Hotplug-VF-Device.sh:44]: Is this infinitive loop?

File: Testscripts/Linux/SRIOV-Hotplug-VF-Device.sh [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+EOF
+
+bash Start_Iperf_Client.sh &
+
+for ((counter=1; counter<10; ++counter))
+do
+    LogMsg "The $counter iterations to disable and enable the VF device"
+    # Disable the VF
+    vf_pci_address=$(lspci | grep -i Ethernet | awk '{ print $1 }')
+    echo 1 >  /sys/bus/pci/devices/${vf_pci_address}/remove
```
Comment (juhlee-microsoft) [inline Testscripts/Linux/SRIOV-Hotplug-VF-Device.sh]: it is nice to check out the path before execution. Agree?

File: Testscripts/Linux/SRIOV-Hotplug-VF-Device.sh [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    sleep 5
+
+    # Using the lspci command, verify if NIC has SR-IOV support
+    lspci -vvv | grep 'mlx4_core\|mlx4_en\|ixgbevf'
```
Comment (juhlee-microsoft) [inline Testscripts/Linux/SRIOV-Hotplug-VF-Device.sh:68]: is this mlx4 only?

File: Testscripts/Linux/SRIOV-Hotplug-VF-Device.sh [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    vf_interface=$(ls /sys/class/net/ | grep -v 'eth0\|eth1\|lo' | head -1)
+else
+    synthetic_interface=$(ip addr | grep "$VF_IP1" | awk '{print $NF}')
+    vf_interface=$(find /sys/devices/* -name "*${synthetic_interface}" | grep "pci" | sed 's/\// /g' | awk '{print $12}')
+fi
+
+# Check the VF is still working
+tx_value=$(cat /sys/class/net/"${vf_interface}"/statistics/tx_packets)
+sleep 4
+tx_value_current=$(cat /sys/class/net/"${vf_interface}"/statistics/tx_packets)
```
Comment (juhlee-microsoft) [inline Testscripts/Linux/SRIOV-Hotplug-VF-Device.sh:96]: Check out the path if exists. (This is test script so we expect any failure)

File: Testscripts/Linux/SRIOV-Hotplug-VF-Device.sh [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+diff=$(expr $tx_value_current - $tx_value)
+if [ "$diff" -lt 50000 ]; then
```
Comment (juhlee-microsoft) [inline Testscripts/Linux/SRIOV-Hotplug-VF-Device.sh:98]: how do you get this number 50K?

**Review by juhlee-microsoft — DISMISSED** [review]

File: Testscripts/Linux/SRIOV-Hotplug-VF-Device.sh (hunk 4 above)
Comment (chvalean) [reply Testscripts/Linux/SRIOV-Hotplug-VF-Device.sh:96]: this path should be LSB, no need to check

Comment (juhlee-microsoft) [reply Testscripts/Linux/SRIOV-Hotplug-VF-Device.sh:96]: Ok, if this is guaranteed 100%, then I close this comment

File: Testscripts/Linux/SRIOV-Hotplug-VF-Device.sh (hunk 1 above)
Comment (xian123) [reply Testscripts/Linux/SRIOV-Hotplug-VF-Device.sh:44]: Yes. Keep running traffic between two VMs in background.

File: Testscripts/Linux/SRIOV-Hotplug-VF-Device.sh (hunk 2 above)
Comment (xian123) [reply Testscripts/Linux/SRIOV-Hotplug-VF-Device.sh]: Adopted.

File: Testscripts/Linux/SRIOV-Hotplug-VF-Device.sh (hunk 3 above)
Comment (xian123) [reply Testscripts/Linux/SRIOV-Hotplug-VF-Device.sh:68]: I only know the mlx4 is supported in azure, if other drivers are supported. Please let me know and I'll update it.

File: Testscripts/Linux/SRIOV-Hotplug-VF-Device.sh (hunk 5 above)
Comment (xian123) [reply Testscripts/Linux/SRIOV-Hotplug-VF-Device.sh:98]: If there is no traffic between 2 VMs, the packet is less than 5k. IMO, 50K is enough though the real packets are bigger than 50k if the VF is working.

**Comment by xian123** [comment]
@chvalean The test supports Ubuntu, RHEL 7.x and SLES15.

**Review by chvalean — APPROVED** [review]

**Pull request merged**